lib/reservation.rb

```diff
@@ -0,0 +1,47 @@
+# Require gems
+require "date"
+
+# Require relatives
+require_relative "room.rb"
+
+module Hotel
+  class Reservation
+    attr_reader :check_in, :check_out, :room, :total_cost, :id
+
+    def initialize(check_in:, check_out:, room:, total_cost: nil, id: nil)
+      @check_in = Date.parse(check_in)
+      @check_out = Date.parse(check_out)
+      @room = room
+      @total_cost = total_cost(check_in, check_out, room.cost_per_night)
+      @id = id
+
+      validate_date
+    end
+
+    def total_cost(check_in, check_out, cost_per_night)
+      number_of_nights = number_of_nights(check_in, check_out)
+      total_cost = cost_per_night * number_of_nights
+
```

Comment:
I like the decision to make this a method rather than storing the cost as an instance variable. That way if the number of nights or the cost per night were to change (say in a future version of the program), the total cost will automatically be correct.